binance_trade_bot/database_warmup.py

```diff
@@ -0,0 +1,147 @@
+from .binance_api_manager import BinanceAPIManager
+from .auto_trader import AutoTrader
+from .models.coin import Coin
+from .models.pair import Pair
+
+class WarmUpManager(BinanceAPIManager):
+    def get_all_symbol_tickers(self):       
+        return self.binance_client.get_symbol_ticker()
+
+class WarmUpDatabase(Database):
+    def __init__(self, logger: Logger, config: Config, uri="sqlite:///data/crypto_trading.db"):
+        super().__init__(logger, config, uri)
+
+    def set_coins_to_warmup(self, symbols: List[str], warmup_symbols: List[str]):
+        session: Session
+
+         # Add coins to the database and set them as enabled or not
+        with self.db_session() as session:
+            coins: List[Coin] = session.query(Coin).all()
+
+            # For all warmup symbols and add them to the database if they don't exist
+            for symbol in warmup_symbols:
+                coin = next((coin for coin in coins if coin.symbol == symbol), None)
+                if coin is None:
+                    newCoin = Coin(symbol, False)
+                    session.add(newCoin)
+                    coins.append(newCoin)
+
+            # For all the coins in the database, if the symbol no longer appears
+            # in the config file, set the coin as disabled
+            for coin in coins:
+                if coin.symbol not in warmup_symbols:
+                    coin.enabled = False
+
+            # For all the symbols in the config file, add them to the database
+            # if they don't exist
+            for symbol in symbols:
+                coin = next((coin for coin in coins if coin.symbol == symbol), None)
+                if coin is None:
+                    session.add(Coin(symbol))
+                else:
+                    coin.enabled = True
+
+        # For all the combinations of coins in the database, add a pair to the database
+        with self.db_session() as session:
+            c1 = aliased(Coin)
+            c2 = aliased(Coin)
+            p = aliased(Pair)
+
+            #select all pairs with non exiting pair entry in db and add the missing pair entries
+            query = session.query(c1, c2).\
+                join(c2, c2.symbol != c1.symbol).\
+                outerjoin(p, p.from_coin_id == c1.symbol and p.to_coin_id == c2.symbol).\
+                filter(p.id == None)
+
+            pairs = query.all()
+            for fromCoin, toCoin in pairs:
+                #exclude bridge coin pairs
+                if fromCoin.symbol != self.config.BRIDGE_SYMBOL and toCoin.symbol != self.config.BRIDGE_SYMBOL:
+                    session.add(Pair(fromCoin, toCoin))
+
+class WarmUpTrader(AutoTrader):
+
+    def initialize_trade_thresholds(self):
+        """
+        Initialize the buying threshold of all the coins for trading between them
+        """
+        session: Session
+        with self.db.db_session() as session:
+            for pair in session.query(Pair).filter(Pair.ratio.is_(None)).all():
```

Comment:
@shanejones The warm up script does exactly that. It generates ratios for all coins to all coins where no ratio is given. Not just for the current coin.